File: android/recording/src/main/kotlin/app/recly/recording/SilenceMonitor.kt
package app.recly.recording

import android.media.AudioManager
import android.media.AudioRecordingConfiguration
import android.os.Handler
import android.os.Looper
import recly.core.model.Range

/**
 * The `silenced` ranges of docs/03: stretches where the platform kept the session alive but fed it
 * zeroes (`isClientSilenced` — a privacy toggle, a higher-priority capture, a call). They are not
 * gaps: the audio is there, it is just silent, and the meta has to say which part is which.
 *
 * The transitions arrive on a system callback and the ranges are written into the meta once, at
 * stop. Only [onSilenced] carries logic, so the merging is testable without a device.
 */
class SilenceMonitor(private val elapsedSec: () -> Double) {

    private val ranges = mutableListOf<Range>()
    private var openStartSec: Double? = null
    private var silenced = false
    private var registered: AudioManager.AudioRecordingCallback? = null

    /**
     * Since Android 10 an ordinary app only ever sees its own configurations here, and
     * `MediaRecorder` exposes no session id to filter on, so every reported config is ours.
     */
    @Synchronized
    fun start(audioManager: AudioManager) {
        val callback = object : AudioManager.AudioRecordingCallback() {
            override fun onRecordingConfigChanged(configs: MutableList<AudioRecordingConfiguration>) {
                synchronized(this@SilenceMonitor) {
                    if (registered !== this) return
                    onSilenced(configs.any { it.isClientSilenced }, elapsedSec())
                }
            }
        }
        registered = callback
        audioManager.registerAudioRecordingCallback(callback, Handler(Looper.getMainLooper()))
    }

    /** Unregisters, closes a range still open, and returns everything for the meta. */
    @Synchronized
    fun stop(audioManager: AudioManager): List<Range> {
        val callback = registered
        registered = null
        callback?.let { runCatching { audioManager.unregisterAudioRecordingCallback(it) } }
        onSilenced(false, elapsedSec())
        return ranges.toList()
    }

    /**
     * Only transitions count: the callback fires on every routing change, and a run of "still
     * silenced" must stay one range. A range that resumes where the previous one ended (a flap the
     * callback reported twice) is merged into it rather than appended.
     */
    @Synchronized
    internal fun onSilenced(nowSilenced: Boolean, atSec: Double) {
        if (nowSilenced == silenced) return
        silenced = nowSilenced
        if (nowSilenced) {
            openStartSec = atSec
            return
        }
        val startSec = openStartSec ?: return
        openStartSec = null
        if (atSec <= startSec) return
        val previous = ranges.lastOrNull()
        if (previous != null && startSec <= previous.endSec) {
            ranges[ranges.lastIndex] = previous.copy(endSec = atSec)
        } else {
            ranges += Range(startSec = startSec, endSec = atSec, reason = REASON)
        }
    }

    @Synchronized
    internal fun ranges(): List<Range> = ranges.toList()

    private companion object {
        /** docs/03's example reason for a lost microphone. */
        const val REASON = "mic_taken"
    }
}

File: apple/RecKit/Sources/RecKit/Recorder/SilenceMonitor.swift
import Foundation
import ReclyCore

/// The `silenced` ranges of docs/03 on Apple: the stretches where something else held the
/// microphone — an iOS interruption, a call or Siri — and the recording carried on afterwards.
/// (The Android `SilenceMonitor` is the same contract over `isClientSilenced`, down to the reason
/// string; the two are meant to stay readable side by side.)
///
/// Two clocks, because an interruption stops the engine: the *position* is the recording's own —
/// frames written, which do not advance while the microphone is gone — and the *length* is the wall
/// clock the interruption lasted. That is the same shape a restart's `gaps` entry is written with
/// (`SegmentedRecorder.restart`), and it says the one thing a reader wants: at this point in the
/// file, this much was taken away.
struct SilenceMonitor {
    /// docs/03's example reason for a lost microphone, and the one Android writes.
    static let reason = "mic_taken"

    private(set) var ranges: [ReclyCore.Range] = []
    /// Where the open range starts, in both clocks. `nil` when the microphone is ours.
    private var openedAt: (positionSec: Double, uptimeSec: Double)?
    /// The recording's position when the last range was closed. It is what says whether any audio
    /// was written since — the range's own `endSec` cannot, because it carries the wall clock the
    /// interruption lasted on top of that position.
    private var closedAtPositionSec: Double?

    /// Only transitions count: a run of "still silenced" is one range, and a second `began` for an
    /// interruption already open has nothing to add.
    ///
    /// A range that resumes where the previous one ended — a flap reported twice — is merged into
    /// it rather than appended, so a call that Siri interrupted is not two holes.
    mutating func set(_ silenced: Bool, positionSec: Double, uptimeSec: Double) {
        guard silenced != (openedAt != nil) else { return }
        guard silenced else { return close(positionSec: positionSec, uptimeSec: uptimeSec) }
        openedAt = (positionSec, uptimeSec)
    }

    /// Closes a range that is still open — a stop, or a recording that ended during the call that
    /// silenced it. Nothing to do when the microphone was ours all along.
    mutating func close(positionSec: Double, uptimeSec: Double) {
        guard let opened = openedAt else { return }
        openedAt = nil
        let lastedSec = max(0, uptimeSec - opened.uptimeSec)
        guard lastedSec > 0 else { return }
        // No frames were written since the last one closed — a call that Siri interrupted, one
        // hole reported twice — so they are one range as long as both put together, not one that
        // ends where the second did. A position that has moved means audio in between: two ranges.
        if let previous = ranges.last, let closedAt = closedAtPositionSec, opened.positionSec <= closedAt {
            ranges[ranges.count - 1] = ReclyCore.Range(
                startSec: previous.startSec,
                endSec: previous.endSec + lastedSec,
                reason: previous.reason
            )
            closedAtPositionSec = opened.positionSec
            return
        }
        ranges.append(
            ReclyCore.Range(
                startSec: opened.positionSec,
                endSec: opened.positionSec + lastedSec,
                reason: Self.reason
            )
        )
        closedAtPositionSec = opened.positionSec
    }
}
